[JSC] Node's JSTokenLocation sometimes doesn't point to the start of an expression

shvaikalesh opened the pull request:
#### e0132b80a4223be92a41f2b65a34b6342eab6f76
<pre>
[JSC] Node&apos;s JSTokenLocation sometimes doesn&apos;t point to the start of an expression
<a href="https://bugs.webkit.org/show_bug.cgi?id=267728">https://bugs.webkit.org/show_bug.cgi?id=267728</a>
&lt;<a href="https://rdar.apple.com/problem/121215280">rdar://problem/121215280</a>&gt;

Reviewed by Justin Michaud.

Before this patch, if we tried to reparse at node-&gt;position(), we would fail with SyntaxError for
object / array literals, async (arrow) functions, and member expressions with `new` / optional chaining,
due to Node&apos;s JSTokenLocation not pointing to the start of an expression.

This change fixes JSTokenLocation to be correct for above-mentioned cases, enabling us to redesign
class field initializers reparsing, and also improves accuracy of expression divots by removing
lastTokenEndPosition() usage from OPENBRACKET / DOT / BACKQUOTE cases of parseMemberExpression().

lastTokenEndPosition() makes sense in case of e.g. parseArguments() because after it exits, the
current token is the next one after `)`, yet in the e.g. DOT case, lastTokenEndPosition() most
likely points to previous identifier, which might be even on a different line.

On top of that, this patch improves stack trace accuracy for `new new new C` expressions by
storing position of each NEW token in a vector.

Aligns stack traces (error locations) with V8 and SpiderMonkey, including for modified API test.

* JSTests/stress/regress-267728.js: Added.
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createNewExpr):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::parseAssignmentExpression):
(JSC::Parser&lt;LexerType&gt;::parseObjectLiteral):
(JSC::Parser&lt;LexerType&gt;::parseArrayLiteral):
(JSC::Parser&lt;LexerType&gt;::parseAsyncFunctionExpression):
(JSC::Parser&lt;LexerType&gt;::parsePrimaryExpression):
(JSC::Parser&lt;LexerType&gt;::parseMemberExpression):
(JSC::Parser&lt;LexerType&gt;::parseArrowFunctionExpression):
* Source/JavaScriptCore/parser/Parser.h:
* Source/JavaScriptCore/parser/SyntaxChecker.h:
(JSC::SyntaxChecker::createNewExpr):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewConfiguration.mm:

Canonical link: <a href="https://commits.webkit.org/273255@main">https://commits.webkit.org/273255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd3065e287e7f5590cc5818213b954dd1ffd438f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13696 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37557 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31452 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36004 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10776 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30394 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11602 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31041 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10143 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10208 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31127 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38812 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/29571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31664 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31455 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36246 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/34701 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8225 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34220 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12135 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/41335 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10857 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4484 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
I think you mean line, not like in the commit message.

r=me, but let's wait for https://github.com/WebKit/WebKit/pull/22880 to land first.

Commit:
[JSC] Node's JSTokenLocation sometimes doesn't point to the start of an expression
https://bugs.webkit.org/show_bug.cgi?id=267728
<rdar://problem/121215280>

Reviewed by Justin Michaud.

Before this patch, if we tried to reparse at node->position(), we would fail with SyntaxError for
object / array literals, async (arrow) functions, and member expressions with `new` / optional chaining,
due to Node's JSTokenLocation not pointing to the start of an expression.

This change fixes JSTokenLocation to be correct for above-mentioned cases, enabling us to redesign
class field initializers reparsing, and also improves accuracy of expression divots by removing
lastTokenEndPosition() usage from OPENBRACKET / DOT / BACKQUOTE cases of parseMemberExpression().

lastTokenEndPosition() makes sense in case of e.g. parseArguments() because after it exits, the
current token is the next one after `)`, yet in the e.g. DOT case, lastTokenEndPosition() most
likely points to previous identifier, which might be even on a different line.

On top of that, this patch improves stack trace accuracy for `new new new C` expressions by
storing position of each NEW token in a vector.

Aligns stack traces (error locations) with V8 and SpiderMonkey, including for modified API test.

* JSTests/stress/regress-267728.js: Added.
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createNewExpr):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::parseAssignmentExpression):
(JSC::Parser<LexerType>::parseObjectLiteral):
(JSC::Parser<LexerType>::parseArrayLiteral):
(JSC::Parser<LexerType>::parseAsyncFunctionExpression):
(JSC::Parser<LexerType>::parsePrimaryExpression):
(JSC::Parser<LexerType>::parseMemberExpression):
(JSC::Parser<LexerType>::parseArrowFunctionExpression):
* Source/JavaScriptCore/parser/Parser.h:
* Source/JavaScriptCore/parser/SyntaxChecker.h:
(JSC::SyntaxChecker::createNewExpr):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewConfiguration.mm:

Canonical link: https://commits.webkit.org/273255@main
